src/transforms/dedupe.rs

```diff
@@ -0,0 +1,414 @@
+use super::Transform;
+use crate::{
+    event::{Event, Value},
+    topology::config::{DataType, TransformConfig, TransformContext, TransformDescription},
+};
+use bytes::Bytes;
+use chrono::{DateTime, Utc};
+use lru::LruCache;
+use serde::{Deserialize, Serialize};
+use std::any::TypeId;
+use std::collections::BTreeMap;
+use string_cache::DefaultAtom as Atom;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub enum FieldMatchConfig {
+    #[serde(rename = "match")]
+    MatchFields(Vec<Atom>),
+    #[serde(rename = "ignore")]
+    IgnoreFields(Vec<Atom>),
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct CacheConfig {
+    pub num_entries: usize,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct DedupeConfig {
+    #[serde(default = "default_field_match_config")]
+    pub fields: FieldMatchConfig,
+    #[serde(default = "default_cache_config")]
+    pub cache: CacheConfig,
+}
+
+fn default_cache_config() -> CacheConfig {
+    CacheConfig { num_entries: 5000 }
+}
+
+fn default_field_match_config() -> FieldMatchConfig {
+    FieldMatchConfig::MatchFields(vec!["timestamp".into()])
+}
+
+pub struct Dedupe {
+    config: DedupeConfig,
+    cache: LruCache<CacheEntry, bool>,
+}
+
+inventory::submit! {
+    TransformDescription::new_without_default::<DedupeConfig>("dedupe")
+}
+
+#[typetag::serde(name = "dedupe")]
+impl TransformConfig for DedupeConfig {
+    fn build(&self, _cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        Ok(Box::new(Dedupe::new(
+            self.cache.num_entries,
+            self.fields.clone(),
+        )))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "dedupe"
+    }
+}
+
+/**
+* A CacheEntry comes in two forms, depending on the FieldMatchConfig in use.
+*
+* When matching fields, a CacheEntry contains a vector of optional 2-tuples.  Each element in the
+* vector represents one field in the corresponding LogEvent.  Elements in the vector will correspond
+* 1:1 (and in order) to the fields specified in "fields.match".  The tuples each store the
```

Comment:
This is really interesting to me. Do we consider `Event { a: "foo", b: "bar" }` to be different than `Event { b: "bar", a: "foo" }`?

Reply:
I see the `dedupe_match_field_order_irrelevant` test and wonder if this comment is still applicable?

Reply:
This patch certainly considers them the same.  Given the recent work to change how we store Events to be backed by BTreeMap instead of HashMap, which effectively coerces all events to the same field order, that seemed like the desired behavior.

Reply:
> I see the `dedupe_match_field_order_irrelevant` test and wonder if this comment is still applicable?

Not sure I follow.  This comment is explaining how the code ensures that the field order doesn't matter, and thus why that test passes.